nexus_constructor/instrument_view/q_components.py

```diff
@@ -0,0 +1,168 @@
+from typing import List
+
+import numpy as np
+from PySide2.Qt3DCore import Qt3DCore
+from PySide2.Qt3DExtras import Qt3DExtras
+from PySide2.Qt3DRender import Qt3DRender
+from PySide2.QtGui import QColor, QMatrix4x4, QVector3D
+
+from nexus_constructor.instrument_view.off_renderer import OffMesh
+from nexus_constructor.instrument_view.qentity_utils import (
+    MATERIAL_ALPHA,
+    MATERIAL_COLORS,
+    MATERIAL_DIFFUSE_COLORS,
+    create_material,
+    create_qentity,
+)
+
+
+class QComponent:
+    def __init__(self, root_entity: Qt3DCore.QEntity, nx_class: str):
+        self.root_entity = root_entity
+        self.nx_class = nx_class
+
+        self.entities: List[Qt3DCore.QEntity] = []
+        self.default_material: Qt3DRender.QMaterial = self._create_default_material()
+
+    def create_entities(self):
+        raise NotImplementedError
+
+    def add_transformation(self, transformation: Qt3DCore.QComponent):
+        raise NotImplementedError
+
+    def remove_transformation(self, transformation: Qt3DCore.QComponent):
+        for entity in self.entities:
+            entity.removeComponent(transformation)
+
+    def setParent(self, value=None):
+        for entity in self.entities:
+            entity.setParent(value)
+
+    def get_entity(self):
+        return self.entities[0]
+
+    def _create_default_material(self) -> Qt3DRender.QMaterial:
+        return create_material(
+            MATERIAL_COLORS.get(self.nx_class, QColor("black")),
+            MATERIAL_DIFFUSE_COLORS.get(self.nx_class, QColor("grey")),
+            self.root_entity,
+            MATERIAL_ALPHA.get(self.nx_class),
+        )
+
+
+class OffMeshQComponent(QComponent):
+    def __init__(self, mesh: OffMesh, root_entity: Qt3DCore.QEntity, nx_class: str):
+        super().__init__(root_entity, nx_class)
+        self._mesh = mesh
+
+    def create_entities(self):
+        self.entities.append(
+            create_qentity([self._mesh, self.default_material], self.root_entity)
+        )
+
+    def add_transformation(self, transformation: Qt3DCore.QComponent):
+        self.entities[0].addComponent(transformation)
```

Comment:
also here maybe add a short docstring. Does it apply transformation to all entities?

Reply:
Ah, it only has one component in the list? it might still be good to go through the list and add the transformation for all entities in the list even if we are pretty sure it will only contain one for all future.